### DISCUSSION TITLE
simple prototype for tvp

### DISCUSSION THREAD

**Discussion by DavidePatria** [do-mpc/do-mpc#145] [Q&A]
I would need to use a time varying parameter for control testing but I cannot figure out how to use the function.
the tvp would be variable to be put in the objective function in order to vary the final position of the vehicle I'm considering.
As a start knowing how to make it have different values based on the time instant would make the job.
I understand that the functions are defined as functions of time, therefore some nested if-else statement would be enough, so this is not the main problem, but rather how to use a tvp in the problem, and specifically in the mterm.
thank you

**Comment by 4flixt** [comment]
Hi Davide,

Using the tvp variables in the objective function is very easy. After you have defined it in the model:

```python
state = model.set_variable('_x', 'state')
setpoint = model.set_variable('_tvp', 'setpoint')
```
You can just use this anywhere in your MPC configuration:
```python
mterm = (state-setpoint)**2
mpc.set_objective(mterm, lterm)
```
At this point it is simply a symbolic variable for each step of the prediction horizon. 
The difficult part is to set the right NUMERICAL values for the right time-steps when computing the MPC solution.
This has to be done via the ``tvp_fun``. I think you'll find most of the information [here](https://www.do-mpc.com/en/latest/FAQ.html#time-varying-parameters)

Best
4flixt

**Reply by DavidePatria** [reply]
I don't quite understand if the for loop to set the tvp for many time steps is mandatory or not. 
I intend to use it only in the MPC controller cost function, do I still need to define it in the simulator>?

I have defined my function like so 
```
model = do_mpc.model.Model(model_type)

alt_fin = model.set_variable(var_type='_tvp', var_name='alt_fin')
model.setup()

tvp_template = mpc.get_tvp_template()

def tvp_fun(t_now):
    if t_now < 1000:
        for k in range(n_horizon+1):
            tvp_template['_tvp',k,'alt_fin'] = 4 
    else:
        for k in range(n_horizon+1):
            tvp_template['_tvp',k,'alt_fin'] = 1 
    return tvp_template

mpc.set_tvp_fun(tvp_fun)
mpc.setup()

```

if it isn't defined in simulator through `simulator.set_tvp_fun(tvp_fun)` I get:
 _Exception: You have not supplied a function to obtain the time-varying parameters defined in model. Use .set_tvp_fun() prior to setup._

even if I define it in the simulator I still get: 
```
--> 356 simulator.set_tvp_fun(tvp_fun)
AssertionError: Incorrect output of tvp_fun. Use get_tvp_template to obtain the required structure.
```

what am I doing wrong? how am I supposed to define this for the sole purpose of making the cost function vary over time?

**Reply by Filippo-Tamagnini** [reply]
Perhaps i can suggest a dirty solution to this problem, I'm not sure if it's the intended way of doing things but here it goes: the tvp_fun you must provide is a python function, which can depend on externally defined variables. I.e: 

```
val = 3
def get_val():
    return [1, val, val**2]
print(get_val())
val = 4
print(get_val())
```
As you can see from the example, the output of the function depends on the value of the variable val (which exists in the global namespace), and not on the value that such variable had during the definition of the function.

**Reply by DavidePatria** [reply]
actually I got it working by defining an empty function for the simulator and one that sets the tvp as an array with a length equal to the n_horizon+1 as shown in the faq, but I'm still missing the dependency on time

the function I wrote is this:
```
def tvp_functio(t_now):
    if t_now < 50:
        for k in range(n_horizon+1):
            tvp_bozza['_tvp',k,'alt_fin'] = 4
    else:
        for k in range(n_horizon+1):
            tvp_bozza['_tvp',k,'alt_fin'] = 1
    return tvp_bozza
```
where alt_fin is the value I want to change during the execution, which is present in the cost function.
given that my simulation lasts 1000 instants at 0,01 seconds and that the mpc has a sample time of 0,1 seconds instead, the number of instants for the mpc 1/10th, as shown by the size of the array in mpc.data['_x','whatever'].
This is why I put `if t_now < 50:`, supponsing that t_now is the number of the present time instant, but this doesn't work as expected.
By modifying the values I found that the number in the if statement relates the simulation instants as t_if / t_step_mpc
In my case the step time for the MPC is set to 0,2 (very demanding task), and by using `if t_now < 5:` I get 2,5 seconds as a threshold.

The formula seems to be (0,1*t_if) / t_step_mpc = t_sim [s]. at first I tough the t_now thresholds in the tvp_fun would be related to the simulator time instants (in my case 0,01 of time step) but, more logically, these are related to the mpc time step.
can I get a confirmation on this?

**Reply by Filippo-Tamagnini** [reply]
Ah, I see your point now! Yes, the time step of the tvp_fun supplied to the controller should correspond to the controller time step.
@4flixt Imo, the way tvps are implemented is really confusing. I'm sure there's some logic behind this implementation, but from the user perspective two functions with the same name should share a reasonably similar behaviour (and possibly also be directly reusable).
Moreover, the fact that the tvp_fun is dependent on time should be sufficient in and of itself to define the value of the parameters at any point in time, independently from the choice of the step parameters of the controller.

**Comment by DavidePatria** [comment]
I set the controller to 0,1 sec of execution and internal time step, now the if statement in the tvp function can be set directly with seconds. so
`if t_now > 5 ` refers to seconds.

what gives me an headache is thinking what I would need to do to implement a more complex function, like a sine, for the tvp parameter, having to set its value for the entire horizon every iteration.
of course I can come up with an idea, and I'll probably implement it, but I find this a bit cumbersome to be honest.
would it be enough to define a tvp parameter as an array of value referred to the simulator time? using this sampling time would allow to sample the array at the required frequency, so changing the time constant of the controller (sampling time and execution time) wouldn't change much.

anyhow, I think once the implementation of a sine will be done I might contribute with this example in the FAQ if I'm allowed to. I understood somet things that I think are worth being written somewhere to speed up the use of this library.

EDIT: at this point I'm becoming pedantic, I know, handle this infos as you wish 
in the function of my tvp I use
```
    if t_now >= 10 and t_now < 20: 
        for k in range(n_horizon+1):
            tvp_template['_tvp',k,'alt_fin'] = 4 
```
and the condition is verified only when the simulation time is 10,1 seconds. If I want the the if to be verified exactly at 10s i need to set the lower threshold at 9.9 (still, the t_step of the controller is 0,1) and this works.

I have checked by printing t_now from inside the tvp_fun, which I believe is called in the simulation loop when `u0 = mpc.make_step(x0)` is called. 
By doing this and checking other things I can confirm that the condition `if t_now >= 10` doesn't get verified for `t_sim = 10,0` but only at `t_sim = 9,9`. For now I just manually set this to 9.9, but there isn't supposed to be a discrepancy given that both arrays start from 0 and the values match, maybe I lack some basic knowledge of python function, which is likely at this point.
